server/src/main/java/org/elasticsearch/index/engine/InternalEngine.java

```diff
@@ -2177,7 +2178,8 @@ private IndexWriterConfig getIndexWriterConfig() {
         if (softDeleteEnabled) {
             mergePolicy = new RecoverySourcePruneMergePolicy(SourceFieldMapper.RECOVERY_SOURCE_NAME, softDeletesPolicy::getRetentionQuery,
                 new SoftDeletesRetentionMergePolicy(Lucene.SOFT_DELETES_FIELD, softDeletesPolicy::getRetentionQuery,
-                    new PrunePostingsMergePolicy(mergePolicy, IdFieldMapper.NAME)));
+                    new PrunePostingsMergePolicy(mergePolicy, IdFieldMapper.NAME, () -> LongPoint.newRangeQuery(
```

Comment:
I wonder if passing in the actual seqNo instead could allow the prune id policy to pre-check if the segment has any seqNos above the localCheckPoint (by checking the max-seqNo of the segment). The speculation here is that executing a query has a relatively large overhead compared to that check, but I am into unknown territory and this may be perfectly fine as is if that overhead is negligible.

Reply:
@jpountz  raised a similar optimization before. Yes, we can explore this.